### PR TITLE
Element-wise min and max operations.

### PR DESCRIPTION
**Pull request by dominikgrewe**
Add cmin & cmax operation for element-wise min and max between two
tensors and between a tensor and a scalar.

### CODE REVIEW

**Comment by dominikgrewe** [comment]
CUDA implementations at torch/cutorch#198

File: lib/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -1066,6 +1066,30 @@ void THTensor_(cross)(THTensor *r_, THTensor *a, THTensor *b, int dimension)
                        r__data[2*r__stride] = a_data[0*a_stride]*b_data[1*b_stride] - a_data[1*a_stride]*b_data[0*b_stride];);
 }
 
+void THTensor_(cmax)(THTensor *r, THTensor *t, THTensor *src) {
+  THTensor_(resizeAs)(r, t);
+  TH_TENSOR_APPLY3(real, r, real, t, real, src,
```
Comment (tkoeppe) [inline lib/TH/generic/THTensorMath.c]: This is Stepanov's Mistake in action (it took Stepanov 15 years to get min and max right):

You get away with this in C because you are comparing a fixed set of types, but the logic is flawed here: Suppose `<` is only a strict weak order which has equivalent elements, and suppose that `a` and `b` are equivalent but non-equal. You really want `min(a, b)` and `max(a, b)` to be two different elements!

The correct way to write these algorithms is to keep the comparison consistent (and parcimonious, using only a single operator):

    max(a, b) = a < b ? b : a
    min(a, b) = a < b ? a : b

In a nutshell, all the ordering operations (min, max, stable partition, nth element, partial sort, etc.) need to be defined consistently with respect to one another.

Comment (dominikgrewe) [inline lib/TH/generic/THTensorMath.c]: This code is consistent with the rest of Torch (see for example the implementations of `minall` and `maxall`). If we wanted to change it, it should be changed everywhere.

**Comment by hughperkins** [comment]
@tkoeppe these are floats.  Can you give an example of two floats which are 'equivalent but not equal'?  I came up with: a pair of nans.  I think infinities are equal in IEEE floats? (certainly, in th:
```
th> a = 1/ 0
                                                                      [0.0001s]
th> b = 2 / 0
                                                                      [0.0001s]
th> a == b
true	
                                                                      [0.0001s]
th> a < b
false	
                                                                      [0.0001s]
th> a > b
false	
                                                                      [0.0001s]
th> b < a
false	
                                                                      [0.0001s]
th> b > a
false	
                                                                      [0.0001s]
th> 
```

and for nan:
```
th> a = 0/0
                                                                      [0.0001s]
th> b = 0/0
                                                                      [0.0001s]
th> a == b
false	
                                                                      [0.0001s]
th> a == a
false	
                                                                      [0.0001s]
th> b == b
false	
                                                                      [0.0001s]
th> a < b
false	
                                                                      [0.0001s]
th> a > b
false	
                                                                      [0.0000s]
th> b < a
false	
                                                                      [0.0001s]
th> b > a
false	
                                                                      [0.0001s]
```
)

**Comment by tkoeppe** [comment]
Floats aren't a problem, which is why I said that you're probably fine in C. The problem is purely conceptual at this point; it is a misunderstanding of what "min" and "max" mean, rather than a miscomputation. The result of the computation happens to be right, but the understanding isn't. If someone takes the same approach and applies it to a different type of data (e.g. indirection through pointers?), they may surprisingly find that their algorithm doesn't actually work.

So no action is strictly required, but I would find emotional value in polishing the details of the high-level concepts in a math-based library :-)

**Comment by dominikgrewe** [comment]
@soumith Any thoughts on this PR?

**Comment by soumith** [comment]
your tests have lots of non-local variables. I think if you fix them, the PR is good to go. (the variables a,b,c,v)

**Comment by dominikgrewe** [comment]
Sorry about that. That's now fixed.

**Comment by soumith** [comment]
Thanks Dominik

**Comment by dominikgrewe** [comment]
Thanks Soumith. There's also an equivalent cutorch PR: torch/cutorch#198
